src/types/ffi_url.rs

```diff
@@ -0,0 +1,57 @@
+use uniffi::deps::bytes::Buf;
+
+use crate::prelude::*;
+use crate::UniFfiTag;
+
+/// A wrapper around `Url` that allows us to safely deal with Urls generated on hosts.
+///
+/// Context: We have defined a custom type conversion between Rust's `Url` and the hosts analogues.
+/// However, a Url could be considered valid on host but not on Rust. For example, Swift allows to build a Url
+/// from string `"invalid input"`, while Rust doesn't.
+///
+/// Therefore, if a given Rust function expects a `Url` as param and is sent one from host side which is invalid,
+/// the code will panic. However, if we send the wrapper instead, we make sure the conversion is safely done on the
+/// host side, dealing with the failing conversion properly rather than panicking.
+#[derive(Debug, PartialEq, Eq, Hash, uniffi::Object, derive_more::Display)]
+#[uniffi::export(Debug, Display, Eq, Hash)]
+pub struct FfiUrl {
+    pub url: Url,
+}
+
+#[uniffi::export]
+impl FfiUrl {
+    #[uniffi::constructor]
+    pub fn new(url: Url) -> Result<Self> {
```

Comment:
Do we really want uniffi intialiser through URL? We are again hiding the information here as to why this constructor is fallible.
Why not initialise from String?

Reply:
I reacted to that too. Right, we should use `String`, and then we **can** use `FromStr` impl in this `#[uniffi::constructor]`.

Reply:
I reacted to that too. Right, we should use `String`, and then we **can** use `FromStr` impl in this `#[uniffi::constructor]`.

Reply:
I don't think we are "hiding" the reason of the failure, we are just exposing -with the wrapper- that the conversion from `Swift.URL` to `Rust.Url` could fail. Initialising from the String seems unnecessary to me, as we are adding the same logic that we define on [lib.rs](https://github.com/radixdlt/sargon/blob/main/src/lib.rs#L263) but in two places. 
By having it as is, we make sure that the wrapper uses the same logic defined for `UniffiCustomTypeConverter`, with the addition of handling the error on host side.

Reply:
This `FfiUrl` should ideally just forward the underlying `Rust.Url` API, so the constructor should be the primary `Rust.Url` constructor.
as for the same logic with the `UniffiCustomTypeConverter`, well you save one function call - `Url::parse`.
IMO, I would opt for clarity here rather than for optimising one line of code.

Reply:
My reasoning was not because of the extra line of code (in fact, it is going to be one less due to changes on Swift side), but rather to keep the conversion logic on just one place.